File: src/ibvs_controller_cpp/src/camera_controller.cpp
#include "ibvs_controller_cpp/camera_controller.hpp"
#include <opencv2/imgproc.hpp>
#include <opencv2/calib3d.hpp>

namespace ibvs_controller
{

CameraController::CameraController(const rclcpp::NodeOptions & options)
: Node("camera_controller", options),
  camera_info_received_(false),
  lambda_gain_(0.1),
  default_depth_(1.5f),
  last_valid_depth_(1.5f)
{
  // Declare parameters
  this->declare_parameter("lambda_gain", 0.1);

  // Get parameters
  lambda_gain_ = this->get_parameter("lambda_gain").as_double();

  // TODO: Students should initialize their chosen feature detector here
  // Example: feature_detector_ = cv::ORB::create();
  // Note: feature_detector_ is declared but not initialized - this must be done.

  // Create subscribers
  image_sub_ = this->create_subscription<sensor_msgs::msg::Image>(
    "/ibvs/image_raw", 10,
    std::bind(&CameraController::imageCallback, this, std::placeholders::_1));

  depth_sub_ = this->create_subscription<sensor_msgs::msg::Image>(
    "/ibvs/depth/image_raw", 10,
    std::bind(&CameraController::depthCallback, this, std::placeholders::_1));

  camera_info_sub_ = this->create_subscription<sensor_msgs::msg::CameraInfo>(
    "/ibvs/camera_info", 10,
    std::bind(&CameraController::cameraInfoCallback, this, std::placeholders::_1));

  // Create publisher
  cmd_vel_pub_ = this->create_publisher<geometry_msgs::msg::Twist>("/ibvs/cmd_vel", 10);

  // Create control timer (10 Hz)
  control_timer_ = this->create_wall_timer(
    std::chrono::milliseconds(100),
    std::bind(&CameraController::controlLoop, this));

  RCLCPP_INFO(this->get_logger(), "Camera Controller (C++) Node Started");
  RCLCPP_INFO(this->get_logger(), "Lambda gain: %.2f", lambda_gain_);
  RCLCPP_WARN(this->get_logger(), "Note: Using depth workaround for Gazebo Sim 8.x depth sensor limitations");
  RCLCPP_WARN(this->get_logger(), "TODO: Students must initialize feature detector!");
}

void CameraController::imageCallback(const sensor_msgs::msg::Image::SharedPtr msg)
{
  try {
    cv_bridge::CvImagePtr cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
    current_image_ = cv_ptr->image;
  } catch (cv_bridge::Exception& e) {
    RCLCPP_ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
  }
}

void CameraController::depthCallback(const sensor_msgs::msg::Image::SharedPtr msg)
{
  try {
    cv_bridge::CvImagePtr cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::TYPE_32FC1);
    cv::Mat depth_raw = cv_ptr->image;

    // Handle Gazebo depth sensor bug: replace -inf/inf with last valid or default depth
    current_depth_ = depth_raw.clone();

    // Count valid depths
    int valid_count = 0;
    float sum_valid = 0;

    for (int i = 0; i < depth_raw.rows; i++) {
      for (int j = 0; j < depth_raw.cols; j++) {
        float depth_value = depth_raw.at<float>(i, j);
        if (std::isfinite(depth_value) && depth_value > 0) {
          valid_count++;
          sum_valid += depth_value;
        } else {
          // Replace invalid with last valid depth
          current_depth_.at<float>(i, j) = last_valid_depth_;
        }
      }
    }

    // Update last valid depth if we have valid readings
    if (valid_count > 0) {
      last_valid_depth_ = sum_valid / valid_count;
    }

    // Log occasionally if many invalid values
    float invalid_percent = 100.0f * (1.0f - float(valid_count) / (depth_raw.rows * depth_raw.cols));
    if (invalid_percent > 10) {
      RCLCPP_DEBUG(this->get_logger(), "Depth: %.1f%% invalid, using %.2fm",
                   invalid_percent, last_valid_depth_);
    }

  } catch (cv_bridge::Exception& e) {
    RCLCPP_ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
  }
}

void CameraController::cameraInfoCallback(const sensor_msgs::msg::CameraInfo::SharedPtr msg)
{
  if (!camera_info_received_) {
    camera_matrix_ = cv::Mat(3, 3, CV_64F);
    for (int i = 0; i < 3; i++) {
      for (int j = 0; j < 3; j++) {
        camera_matrix_.at<double>(i, j) = msg->k[i * 3 + j];
      }
    }
    camera_info_received_ = true;
    RCLCPP_INFO(this->get_logger(), "Camera matrix received");
  }
}

void CameraController::controlLoop()
{
  if (current_image_.empty() || !camera_info_received_) {
    return;
  }

  // Extract features from current image
  auto current_features = extractFeatures(current_image_);

  // Create velocity command
  geometry_msgs::msg::Twist cmd;

  // TODO: Students should implement the actual IBVS control here
  // For now, just publish zero velocity as an example

  // Example implementation structure:
  // 1. Match current features with desired features
  // 2. Compute feature error
  // 3. Compute interaction matrix
  // 4. Apply control law: v = -lambda * L^+ * e

  cmd.linear.x = 0.0;   // Forward/backward
  cmd.linear.y = 0.0;   // Left/right
  cmd.linear.z = 0.0;   // Up/down
  cmd.angular.x = 0.0;  // Roll
  cmd.angular.y = 0.0;  // Pitch
  cmd.angular.z = 0.0;  // Yaw

  // Publish command
  cmd_vel_pub_->publish(cmd);
}

std::vector<cv::KeyPoint> CameraController::extractFeatures(const cv::Mat& image)
{
  std::vector<cv::KeyPoint> keypoints;

  // TODO: Students should implement feature extraction here
  // Steps:
  // 1. Convert image to grayscale if needed
  // 2. Initialize and configure feature detector (e.g., cv::ORB::create())
  // 3. Detect keypoints and compute descriptors
  // 4. Return keypoints for use in control loop

  // Example skeleton (students must complete):
  // cv::Mat gray;
  // if (image.channels() == 3) {
  //   cv::cvtColor(image, gray, cv::COLOR_BGR2GRAY);
  // }
  // feature_detector_->detect(gray, keypoints);

  return keypoints;
}

cv::Mat CameraController::computeInteractionMatrix(
  const std::vector<cv::KeyPoint>& features,
  const cv::Mat& depth_image)
{
  // TODO: Students should implement the interaction matrix computation
  // This is a placeholder that returns an empty matrix

  int num_features = features.size();
  cv::Mat L(2 * num_features, 6, CV_64F, cv::Scalar(0));

  // For each feature, compute its contribution to the interaction matrix
  // L = [ -1/Z   0   x/Z   xy   -(1+x²)   y  ]
  //     [  0   -1/Z  y/Z  1+y²    -xy    -x  ]

  return L;
}

geometry_msgs::msg::Twist CameraController::computeVelocityCommand(
  const std::vector<cv::KeyPoint>& current_features,
  const std::vector<cv::KeyPoint>& desired_features,
  const cv::Mat& interaction_matrix)
{
  geometry_msgs::msg::Twist cmd;

  // TODO: Students should implement the velocity computation
  // v = -lambda * L^+ * e
  // where L^+ is the pseudoinverse of L, and e is the feature error

  return cmd;
}

float CameraController::getDepthAtPixel(int x, int y)
{
  // Return default if no depth image
  if (current_depth_.empty()) {
    return default_depth_;
  }

  // Ensure coordinates are within bounds
  x = std::max(0, std::min(x, current_depth_.cols - 1));
  y = std::max(0, std::min(y, current_depth_.rows - 1));

  float depth_value = current_depth_.at<float>(y, x);

  // Return valid depth or fallback
  return std::isfinite(depth_value) ? depth_value : last_valid_depth_;
}

// TODO: Students can implement helper classes here if needed
// For example: FeatureExtractor, IBVSController, etc.

}  // namespace ibvs_controller

File: src/ibvs_controller/scripts/camera_controller.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image, CameraInfo
from geometry_msgs.msg import Twist
from cv_bridge import CvBridge
import cv2
import numpy as np

class CameraController(Node):
    def __init__(self):
        super().__init__('camera_controller')
        
        # Subscribers
        self.image_sub = self.create_subscription(
            Image, '/ibvs/image_raw', self.image_callback, 10)
        self.depth_sub = self.create_subscription(
            Image, '/ibvs/depth/image_raw', self.depth_callback, 10)
        self.info_sub = self.create_subscription(
            CameraInfo, '/ibvs/camera_info', self.camera_info_callback, 10)
        
        # Publisher for velocity commands
        self.cmd_pub = self.create_publisher(Twist, '/ibvs/cmd_vel', 10)
        
        # CV Bridge for image conversion
        self.bridge = CvBridge()
        
        # Camera parameters
        self.camera_matrix = None
        self.image = None
        self.depth = None
        
        # Control timer (10 Hz)
        self.control_timer = self.create_timer(0.1, self.control_loop)
        
        self.get_logger().info('Camera Controller Node Started')
    
    def camera_info_callback(self, msg):
        if self.camera_matrix is None:
            self.camera_matrix = np.array(msg.k).reshape(3, 3)
            self.get_logger().info('Camera matrix received')
    
    def image_callback(self, msg):
        try:
            self.image = self.bridge.imgmsg_to_cv2(msg, 'bgr8')
        except Exception as e:
            self.get_logger().error(f'Failed to convert image: {e}')
    
    def depth_callback(self, msg):
        try:
            self.depth = self.bridge.imgmsg_to_cv2(msg, 'passthrough')
        except Exception as e:
            self.get_logger().error(f'Failed to convert depth: {e}')
    
    def control_loop(self):
        """Main control loop - students will implement IBVS here"""
        if self.image is None:
            return
        
        # Example: Simple velocity command (students will replace this)
        cmd = Twist()
        
        # TODO: Students implement feature extraction here
        # features = self.extract_features(self.image)
        
        # TODO: Students implement visual servoing control law
        # cmd.linear.x, cmd.linear.y, cmd.angular.z = self.compute_control(features)
        
        # Example movement (remove in student implementation)
        cmd.linear.x = 0.0  # Forward/backward
        cmd.linear.y = 0.0  # Left/right  
        cmd.angular.z = 0.0  # Rotation
        
        self.cmd_pub.publish(cmd)
    
    def extract_features(self, image):
        """Students implement feature extraction (ORB, SIFT, etc.)"""
        # TODO: Implement feature extraction
        pass
    
    def compute_control(self, features, desired_features):
        """Students implement IBVS control law"""
        # TODO: Implement control law
        pass

def main(args=None):
    rclpy.init(args=args)
    controller = CameraController()
    rclpy.spin(controller)
    controller.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
